core/src/main/java/org/apache/metamodel/query/parser/FromItemParser.java

```diff
@@ -136,16 +169,17 @@ private FromItem parseJoinItem(final String itemToken) {
             // rightPart = "b.id"
             final String rightPart = onClause.substring(indexOfEquals + 1).trim();
 
-            leftOn[i] = findSelectItem(leftPart, leftSide, rightSide);
-            rightOn[i] = findSelectItem(rightPart, leftSide, rightSide);
+            leftOn[i] = findSelectItem(leftPart, fromItems.toArray(new FromItem[fromItems.size()]), rightSide);
+            rightOn[i] = findSelectItem(rightPart, fromItems.toArray(new FromItem[fromItems.size()]), rightSide);
```

Comment:
Let's only create that array once and reuse it for both calls to findSelectItem(...)